packages/patrol_cli/lib/src/features/test/android_test_backend.dart

```diff
@@ -60,74 +92,126 @@ class AndroidAppOptions {
   }
 }
 
-class AndroidTestBackend {
+/// Provides functionality to build, install, run, and uninstall Android apps.
+///
+/// This class must be stateless.
+class AndroidTestBackend extends TestBackend {
   AndroidTestBackend({
+    required Adb adb,
     required ProcessManager processManager,
     required Platform platform,
     required FileSystem fs,
     required DisposeScope parentDisposeScope,
     required Logger logger,
-  })  : _processManager = processManager,
+  })  : _adb = adb,
+        _processManager = processManager,
         _fs = fs,
         _platform = platform,
         _disposeScope = DisposeScope(),
         _logger = logger {
     _disposeScope.disposedBy(parentDisposeScope);
   }
 
+  final Adb _adb;
   final ProcessManager _processManager;
   final Platform _platform;
   final FileSystem _fs;
   final DisposeScope _disposeScope;
   final Logger _logger;
 
-  Future<void> run({
-    required Device device,
-    required AndroidAppOptions options,
-  }) async {
-    final targetName = basename(options.target);
-    final task = _logger
-        .task('Building apk for $targetName and running it on ${device.id}');
-
-    final process = await _processManager.start(
-      options.toGradleInvocation(_platform),
-      runInShell: true,
-      environment: {
-        'ANDROID_SERIAL': device.id,
-      },
-      workingDirectory: _fs.currentDirectory.childDirectory('android').path,
-    );
-
-    process
-        .listenStdOut((line) => _logger.detail('\t: $line'))
-        .disposedBy(_disposeScope);
-    process
-        .listenStdErr((line) => _logger.err('\t$line'))
-        .disposedBy(_disposeScope);
+  @override
+  Future<void> build(AndroidAppOptions options, Device device) async {
+    await _disposeScope.run((scope) async {
+      final subject = options.desc;
+      final task = _logger.task('Building $subject');
+
+      Process process;
+      int exitCode;
+
+      // :app:assembleDebug
+
+      process = await _processManager.start(
+        options.toGradleAssembleInvocation(isWindows: _platform.isWindows),
+        runInShell: true,
+        workingDirectory: _fs.currentDirectory.childDirectory('android').path,
+      )
+        ..disposedBy(scope);
+      process.listenStdOut((l) => _logger.detail('\t: $l')).disposedBy(scope);
+      process.listenStdErr((l) => _logger.err('\t$l')).disposedBy(scope);
+      exitCode = await process.exitCode;
+      if (exitCode == 130) {
+        const cause = 'Gradle build interrupted';
+        task.fail('Failed to build $subject ($cause)');
+        throw Exception(cause);
+      } else if (exitCode != 0) {
+        final cause = 'Gradle build failed with code $exitCode';
+        task.fail('Failed to build $subject ($cause)');
+        throw Exception(cause);
+      }
 
-    final exitCode = await process.exitCode;
+      // :app:assembleDebugAndroidTest
```

Comment:
Make those sections private methods maybe?

Reply:
I considered doing this, and even sketched out some private method, but after a short while, I called back to the current state. I'll invoke [rule of 3 again](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming))... – if I had to execute one more `Process`, I'd surely refactor this chunk.